Support secure sockets (WSS)

defeated opened the pull request:
We're using the `--https` option of Middleman 4, but also need to pass the SSL cert & key to the websocket server being run by `middleman-livereload`.

This PR adds support for `:ssl_certificate` & `:ssl_private_key` options (reusing the same names from `middleman-core`) and passes them to the `EventMachine::WebSocket::Connection` arguments in this format:

https://github.com/igrigorik/em-websocket#secure-server

(I confirmed this is "working on my machine", but this plugin doesn't have a test suite, so I  regrettably have not added new tests. I hope that's OK!)

defeated commented:
@tdreyno let me know if you prefer these squashed down, or if there's anything I can do to grease the wheels of progress ;) 🎩 Thanks!

defeated commented:
(I should mention, initially I thought these settings could be inherited from the `app` but those keys in `app.config` return `nil` inside `Middleman::LiveReloadExtension#new`)

sandstrom commented on lib/middleman-livereload/reactor.rb:
```diff
@@ -40,11 +40,25 @@ def reload_browser(paths = [])
 
+      def websocket_options
+        @websocket_options ||= begin
+          return {} unless @options[:ssl_certificate] || @options[:ssl_private_key]
```
Are both required, or just one? If both are required I'd expect `&&` here (or an exception being thrown if only one of them is set).

sandstrom commented on lib/middleman-livereload/reactor.rb:
```diff
@@ -40,11 +40,25 @@ def reload_browser(paths = [])
+          }
+        end
+      end
+
```
How about something like this?

```ruby
def secure_websockets?
  @options[:wss_certificate] && @options[:wss_private_key]
end
```

This method `secure_websockets?` can then be reused in both `websockets_options` and for the logging.

sandstrom commented on lib/middleman-livereload/extension_3_1.rb:
```diff
@@ -13,6 +13,8 @@ class LiveReloadExtension < Extension
     option :livereload_css_target, 'stylesheets/all.css', 'CSS file to load when a @imported CSS partials are modified'
     option :livereload_css_pattern, Regexp.new('_.*\.css'), 'Regexp matching filenames that trigger live reloading target'
+    option :ssl_certificate, nil, 'Path to an X.509 certificate to use for the websocket server'
+    option :ssl_private_key, nil, 'Path to an RSA private key for the websocket certificate'
```
I think `wss_certificate` and `wss_private_key` is better. Both because it's not ssl (it's TLS) and we may want to add encryption for the http-part of middleman-livereload in the future.

sandstrom commented on README.md:
```diff
@@ -58,6 +58,10 @@ To opt out set `livereload_css_target: nil`.
 
+#### `:ssl_certificate` and `:ssl_private_key`
+
+Support secure sockets (WSS) by passing SSL certificate & private key, for
```
SSL -> TLS

defeated commented:
@sandstrom thanks so much for the feedback!

As I made subsequent changes, all the `wss_*` references smelled a bit like "fields not defining state", so I've extracted it all to `Middleman::LiveReloader::Wss`

I also saw `rspec` was included as a dependency, so I added some specs at least around the new `Wss` class behavior.

(And then squashed everything into a single commit)

Let me know what you think?

sandstrom commented on lib/middleman-livereload/wss.rb:
```diff
@@ -0,0 +1,40 @@
+module Middleman
+  module LiveReload
+    class Wss
+      attr_reader :certificate, :private_key
```
Are these methods used outside the class?

If not, can we replace them with private methods? Or just access the instance variables directly within this class (skipping private accessors)

```ruby
private
def certificate
  @certificate
end

# === OR ===

def valid?
  @certificate && @private_key
end

# ... and similar instance variable access in other places too

```

sandstrom commented on lib/middleman-livereload/wss.rb:
```diff
@@ -0,0 +1,40 @@
+      def validate!
+        return unless present?
+        return if valid?
+        raise ArgumentError.new, 'You must provide both :wss_certificate and :wss_private_key'
```
Minor, but I'd go with

```ruby
def validate!
  if present? && !valid?
    raise ArgumentError, 'You must provide both :wss_certificate and :wss_private_key'
  end
end
```

defeated commented:
@sandstrom this should include all feedback now!

sandstrom commented:
@defeated Awesome!

Thanks for spending the time on the adjustments too. Merging! ⛵️

tdreyno commented:
Nice. Thanks @defeated !
